docs: fix warnings for documentation build, use a linter

qmonnet opened the pull request:
Four big steps in the PR:

1. Fix the checks for warnings on builds for the documentation (1st commit).
2. Address build warnings (2nd commit).
3. Introduce a linter, rstcheck, for the RST documentation files (3rd commit).
4. Fix issues reported by linters in the documentation (remaining commits).

Please refer to individual commit logs for details.

qmonnet committed:
docs: fix warnings for documentation build

"txt" is not a valid language to pass to "code-block" directives.
Passing "none" instead of "txt" would work; but there is no need to mark
the snippet as a "code-block" if no syntax highlighting is required,
simply using a literal block is enough.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
docs: report warnings when building documentation

The check-build.sh script would check for warnings produced by running
"sphinx-build -b spelling", but not those produced when generating the
documentation with "sphinx-build -M html".

Let's reuse the warnings.txt file to store these warnings, if any, and
to return with a non-0 exit code in that case.

Notes: The file can be safely truncated (we won't reach that step if
warnings were generated from "sphinx-build -b spelling"). Also we do not
have to care of errors produced by "sphinx-build -M html", as they would
translate in an error exit code that would make the script return before
we look at the warnings.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
docs: pass a linter on the RST files

Use rstcheck (https://github.com/myint/rstcheck/) as a linter for the
RST Documentation files.

Run the linter before syntax and spelling validation by Sphinx, because
it goes much faster, so we fail faster if an error happens at that
stage.

Ignore the following items:

- Custom directives: tabs (from Sphinx extensions).
- Custom roles: any role added with the extlinks extension.
- Language: skip linting for bash in ".. code-block:: bash" snippets.
- Messages:
  - Skip an error on ordinated lists in bpf.rst (we have list with
    decreasing numbers, and there is also a small indent issue in a bit
    enumerated list in the file, but Sphinx handles it well so let's
    leave it untouched).
  - Hyperlink target not referenced: rstcheck does not seem able to
    recognise when targets are referenced from a different file.
  - Duplicate implicit target name: this happens when HTML anchors
    generated from section titles collide with explicit targets (or
    other section titles on the same page). This is harmless and we are
    not looking to fix those.
  - Malformed tables: it would be useful to have these reports if
    rstcheck would support ignoring substitution ("|PATTERN|") in grid
    tables. I reported a bug on rstcheck's tracker.
- Substitutions: any custom substitution defined in conf.py.

Note the existence of an alternative tool, rst-lint
(https://github.com/twolfson/restructuredtext-lint), but from my
experiments it does not integrate well with Sphinx at the moment.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
docs: remove duplicate explicit targets from RST files

The syntax for inline hyperlinks can take one or two underscores at the
end of the marker. With one underscore, it is an explicit target, which
means that the text for the link can be reused elsewhere without having
to copy the URL again. With two underscores, the reference is anonymous
and cannot be reused elsewhere.

Avoid having multiple explicit targets in the documentation. Let's go
for the easiest possible fix: add an additional underscore to make the
targets anonymous.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
docs: indent code blocks in enumerated lists

Following a report from rstcheck, indent the code blocks used in
enumerated list items, to make them part of these items instead of
"breaking" the list.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
docs: fix a few JSON snippets

A JSON snippet in alibabacloud-eni.rst is not valid JSON, there is a
spurious comma after the last element of an array. Let's remove it to
make the JSON valid.

Let's also update the declaration for two blocks that were recently
marked as JSON. The ouptut contained in these blocks is made of JSON
fragments, but each block contains _several_ fragments without a root
object or array, making the entire snippet invalid JSON and trigerring
reports from the linter. Let's simply use literal blocks instead.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

joestringer commented on Documentation/check-build.sh:
```diff
@@ -85,4 +85,11 @@ fi
 
 echo "Building documentation (${target})..."
-exec sphinx-build -M "${target}" "${script_dir}" "${build_dir}" $@
+sphinx-build -M "${target}" "${script_dir}" "${build_dir}" $@ -q 2> >(tee "${warnings}" >&2)
```
I'm curious, what's the difference between having the `exec` at the start vs. not?

qmonnet replied:
I asked the author of the code (Ilya) about that before doing the change, and he answered that it was mostly a “stylistic thing”, that “when a script calls one command at the end and that command is relatively heavyweight, [he] tends to use exec to sort of signify the purpose of the script”.

But I had to remove it because `exec`-ing into the command substitutes that command to the shell, and we wouldn't return to run the new instructions that my commit adds below.

qmonnet committed:
docs: split a C snippet in BPF documentation

A snippet of code was recently marked as C, but it actually contains
both C and a sample output from the kernel verifier (which is not C
code). Sphinx raises a warning as it fails to apply syntax highlighting
for C for that snippet. Let's split the C code from the verifier output.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>
